cmd/admin-trace.go

```diff
@@ -0,0 +1,224 @@
+	console.SetColor("Body", color.New(color.FgYellow))
+	for _, c := range colors {
+		console.SetColor(fmt.Sprintf("Node%d", c), color.New(c))
+	}
+	// Create a new Minio Admin Client
+	client, err := newAdminClient(aliasedURL)
+	if err != nil {
+		fatalIf(err.Trace(aliasedURL), "Cannot initialize admin client.")
+		return nil
+	}
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+
+	// Start listening on all trace activity.
+	traceCh := client.ListenTraceNotification(all, doneCh)
+	for traceInfo := range traceCh {
+		if traceInfo.Err != nil {
+			fatalIf(probe.NewError(traceInfo.Err), "Cannot listen to http trace")
+		}
+		if verbose {
+			printMsg(traceMessage{traceInfo})
+			continue
+		}
+		printMsg(shortTrace(traceInfo))
+	}
+	return nil
+}
+
+// abbeviated trace output
+type traceShort struct {
+	trc.Short
+}
+
+// parse Operation name from function name
+func getOpName(fname string) (op string) {
+	res := funcNameRegex.FindStringSubmatch(fname)
+	op = fname
+	if res != nil && len(res) == 2 {
+		op = res[1]
+	}
+	return
+}
+
+// return a struct with minimal trace info.
+func shortTrace(t trc.Info) traceShort {
+	s := traceShort{}
+	s.NodeName = t.NodeName
+	s.Time = t.ReqInfo.Time
+	s.Host = t.ReqInfo.Host
+	s.FuncName = getOpName(t.FuncName)
+	s.URL = t.ReqInfo.URL
+	s.StatusCode = t.RespInfo.StatusCode
+	s.StatusMsg = t.RespInfo.ErrMsg
+	return s
+}
+func (s traceShort) JSON() string {
+	traceJSONBytes, e := json.MarshalIndent(s, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+	return string(traceJSONBytes)
+}
+func (s traceShort) String() string {
+	var nodeNameStr string
+	if s.NodeName != "" {
+		nodeNameStr = colorizedNodeName(s.NodeName)
+	}
+	msg := fmt.Sprintf("%s %s ", s.Time.Format(timeFormat), console.Colorize("FuncName", s.FuncName))
+	msg += fmt.Sprintf("%s%s ", nodeNameStr, s.URL.Path)
+	msg += console.Colorize("ResponseStatus", fmt.Sprintf("%s %s", http.StatusText(s.StatusCode), s.StatusMsg))
+	return msg
+}
+
+type traceMessage struct {
+	trc.Info
+}
+
+// colorize node name
+func colorizedNodeName(nodeName string) string {
+	nodeHash := fnv.New32a()
+	nodeHash.Write([]byte(nodeName))
+	nHashSum := nodeHash.Sum32()
+	idx := int(nHashSum) % len(colors)
+	var c color.Attribute
+	var i int
+	for i, c = range colors {
+		if idx == i {
+			break
+		}
+	}
+	return console.Colorize(fmt.Sprintf("Node%d", c), fmt.Sprintf("%s", nodeName))
+}
+
+func (t traceMessage) JSON() string {
+	traceMessageJSONBytes, e := json.MarshalIndent(t.Info, "", " ")
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+	return string(traceMessageJSONBytes)
+}
+
+func (t traceMessage) String() string {
+	var nodeNameStr string
+	if t.NodeName != "" {
+		nodeNameStr = fmt.Sprintf("%s ", colorizedNodeName(t.NodeName))
+	}
+	ri := t.ReqInfo
+	rs := t.RespInfo
+	msg := nodeNameStr + console.Colorize("Request", fmt.Sprintf("[REQUEST %s] ", t.FuncName))
+	msg += fmt.Sprintf("[%s] [%s]\n", t.ReqID, ri.Time.Format(timeFormat))
```

Comment:
If this function is called often, it is more efficient to avoid string concatenation with `+` and `+=`. `strings.Builder` provides a way that does not involve much copying.